File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/hey/Hooks.kt
package com.basecamp.hey

import kotlin.time.Duration

/**
 * Observability hooks for the HEY SDK.
 *
 * Two levels of hooks:
 * - **Operation-level**: [onOperationStart]/[onOperationEnd] for semantic SDK operations
 *   (e.g. `Boxes.ListBoxes`).
 * - **Request-level**: [onRequestStart]/[onRequestEnd] for HTTP requests (retries, cache, timing).
 *
 * All callbacks are optional. Implementations should not throw — exceptions are caught and
 * silently ignored to prevent hooks from breaking SDK operations.
 */
interface HeyHooks {
    /** Called when a service operation starts. */
    fun onOperationStart(info: OperationInfo) {}

    /** Called when a service operation completes (success or failure). Always called. */
    fun onOperationEnd(info: OperationInfo, result: OperationResult) {}

    /** Called when an HTTP request starts. Called for each attempt including retries. */
    fun onRequestStart(info: RequestInfo) {}

    /** Called when an HTTP request completes. Called for each attempt including retries. */
    fun onRequestEnd(info: RequestInfo, result: RequestResult) {}

    /** Called before a retry attempt, with the attempt about to be made and the wait before it. */
    fun onRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {}
}

/**
 * Describes a semantic SDK operation: the service and operation as the model names them,
 * the kind of record they touch, and whether they change anything. A hand-written wrapper
 * is free to report itself as something other than the route it sends.
 */
data class OperationInfo(
    /** Logical service (e.g. "Boxes", "TimeTracks"). */
    val service: String,
    /** The operation as the model names it (e.g. "ListBoxes"). */
    val operation: String,
    /** HEY resource type in `snake_case` (e.g. "box", "time_track"). */
    val resourceType: String,
    /** Whether this operation modifies state. */
    val isMutation: Boolean,
    /** Specific resource ID if applicable. */
    val resourceId: Long? = null,
)

/** Result of a service operation. */
data class OperationResult(
    /** Operation duration. */
    val duration: Duration,
    /** Error if the operation failed. */
    val error: Throwable? = null,
)

/** Information about an HTTP request. */
data class RequestInfo(
    /** HTTP method (GET, POST, etc.). */
    val method: String,
    /** Full request URL. */
    val url: String,
    /** Current attempt number (1-based), counted across the whole operation. */
    val attempt: Int,
)

/** Result of an HTTP request. */
data class RequestResult(
    /** HTTP status code (0 if request failed before response). */
    val statusCode: Int,
    /** Request duration. */
    val duration: Duration,
    /** Whether the response was served from cache. */
    val fromCache: Boolean = false,
    /** Error if the request failed, whether or not the SDK went on to resend it. */
    val error: Throwable? = null,
)

/** No-op hooks implementation. Zero overhead when unused. */
object NoopHooks : HeyHooks

/**
 * Combines multiple hooks. Start events fire in order, end events in reverse order
 * (proper nesting of spans/traces).
 */
class ChainHooks(private val hooks: List<HeyHooks>) : HeyHooks {
    override fun onOperationStart(info: OperationInfo) {
        for (h in hooks) runCatching { h.onOperationStart(info) }
    }

    override fun onOperationEnd(info: OperationInfo, result: OperationResult) {
        for (h in hooks.asReversed()) runCatching { h.onOperationEnd(info, result) }
    }

    override fun onRequestStart(info: RequestInfo) {
        for (h in hooks) runCatching { h.onRequestStart(info) }
    }

    override fun onRequestEnd(info: RequestInfo, result: RequestResult) {
        for (h in hooks.asReversed()) runCatching { h.onRequestEnd(info, result) }
    }

    override fun onRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {
        for (h in hooks) runCatching { h.onRetry(info, attempt, error, delayMs) }
    }
}

/** Combines multiple hooks implementations. */
fun chainHooks(vararg hooks: HeyHooks): HeyHooks {
    val active = hooks.filter { it !is NoopHooks }
    return when (active.size) {
        0 -> NoopHooks
        1 -> active[0]
        else -> ChainHooks(active)
    }
}

/** Console logging hooks for debugging and development. */
fun consoleHooks(
    logOperations: Boolean = true,
    logRequests: Boolean = false,
    logRetries: Boolean = true,
): HeyHooks = object : HeyHooks {
    override fun onOperationStart(info: OperationInfo) {
        if (!logOperations) return
        val mutation = if (info.isMutation) " [mutation]" else ""
        val resource = info.resourceId?.let { " #$it" } ?: ""
        println("[HEY] ${info.service}.${info.operation}$resource$mutation")
    }

    override fun onOperationEnd(info: OperationInfo, result: OperationResult) {
        if (!logOperations) return
        if (result.error != null) {
            println("[HEY] ${info.service}.${info.operation} failed (${result.duration}): ${result.error.message}")
        } else {
            println("[HEY] ${info.service}.${info.operation} completed (${result.duration})")
        }
    }

    override fun onRequestStart(info: RequestInfo) {
        if (!logRequests) return
        val retry = if (info.attempt > 1) " (attempt ${info.attempt})" else ""
        println("[HEY] -> ${info.method} ${info.url}$retry")
    }

    override fun onRequestEnd(info: RequestInfo, result: RequestResult) {
        if (!logRequests) return
        val cache = if (result.fromCache) " (cached)" else ""
        println("[HEY] <- ${info.method} ${info.url} ${result.statusCode} (${result.duration})$cache")
    }

    override fun onRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {
        if (!logRetries) return
        println("[HEY] Retrying ${info.method} ${info.url} (attempt $attempt, waiting ${delayMs}ms): ${error.message}")
    }
}

internal fun HeyHooks.safeOperationStart(info: OperationInfo) {
    runCatching { onOperationStart(info) }
}

internal fun HeyHooks.safeOperationEnd(info: OperationInfo, result: OperationResult) {
    runCatching { onOperationEnd(info, result) }
}

internal fun HeyHooks.safeRequestStart(info: RequestInfo) {
    runCatching { onRequestStart(info) }
}

internal fun HeyHooks.safeRequestEnd(info: RequestInfo, result: RequestResult) {
    runCatching { onRequestEnd(info, result) }
}

internal fun HeyHooks.safeRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {
    runCatching { onRetry(info, attempt, error, delayMs) }
}

File: swift/Sources/Hey/Hooks.swift
import Foundation

/// Describes a semantic SDK operation: the service and operation as the model names them, the
/// kind of record they touch, and whether they change anything. A hand-written convenience is
/// free to report itself as something other than the route it sends.
public struct OperationInfo: Sendable, Equatable {
    /// Logical service: `Boxes`, `TimeTracks`.
    public var service: String
    /// The operation as the model names it: `ListBoxes`.
    public var operation: String
    /// HEY resource type in `snake_case`: `box`, `time_track`.
    public var resourceType: String
    /// Whether the operation changes anything.
    public var isMutation: Bool
    /// The record the operation acts on, when it names one.
    public var resourceId: Int?

    public init(service: String, operation: String, resourceType: String, isMutation: Bool, resourceId: Int? = nil) {
        self.service = service
        self.operation = operation
        self.resourceType = resourceType
        self.isMutation = isMutation
        self.resourceId = resourceId
    }
}

/// How an operation ended.
public struct OperationResult: Sendable {
    /// How long the operation took, on a monotonic clock.
    public var duration: Duration
    /// The error it failed with, or nil when it succeeded.
    public var error: (any Error)?

    public init(duration: Duration, error: (any Error)? = nil) {
        self.duration = duration
        self.error = error
    }
}

/// One HTTP request, as the hooks hear of it.
public struct RequestInfo: Sendable, Equatable {
    /// The HTTP method.
    public var method: String
    /// The URL. One that authenticates itself — a storage upload — is cut back to its origin.
    public var url: String
    /// The attempt, from 1, counted across the whole operation.
    public var attempt: Int

    public init(method: String, url: String, attempt: Int) {
        self.method = method
        self.url = url
        self.attempt = attempt
    }
}

/// How one HTTP request ended.
public struct RequestResult: Sendable {
    /// The status HEY answered with, or 0 when there was no answer.
    public var statusCode: Int
    /// How long the request took, on a monotonic clock.
    public var duration: Duration
    /// Whether the answer came out of the response cache.
    public var fromCache: Bool
    /// The error the request failed with, whether or not the client went on to resend it.
    public var error: (any Error)?

    public init(statusCode: Int, duration: Duration, fromCache: Bool = false, error: (any Error)? = nil) {
        self.statusCode = statusCode
        self.duration = duration
        self.fromCache = fromCache
        self.error = error
    }
}

/// Observability hooks, at two levels: an operation (`Boxes.ListBoxes`) and each HTTP request
/// it sends, retries included. Every method has a default that does nothing, so implement only
/// what you need.
///
/// ```swift
/// struct Logging: HeyHooks {
///     func onOperationEnd(_ info: OperationInfo, result: OperationResult) {
///         print("\(info.service).\(info.operation) took \(result.duration)")
///     }
/// }
/// ```
public protocol HeyHooks: Sendable {
    /// An operation starts.
    func onOperationStart(_ info: OperationInfo)
    /// An operation ends, however it ends. Always called for one that started.
    func onOperationEnd(_ info: OperationInfo, result: OperationResult)
    /// A request goes out, including every resend.
    func onRequestStart(_ info: RequestInfo)
    /// A request is answered or fails, including every resend.
    func onRequestEnd(_ info: RequestInfo, result: RequestResult)
    /// A request is about to be sent again: the attempt about to be made, and the wait before it.
    func onRetry(_ info: RequestInfo, attempt: Int, error: any Error, delay: Duration)
}

extension HeyHooks {
    public func onOperationStart(_ info: OperationInfo) {}
    public func onOperationEnd(_ info: OperationInfo, result: OperationResult) {}
    public func onRequestStart(_ info: RequestInfo) {}
    public func onRequestEnd(_ info: RequestInfo, result: RequestResult) {}
    public func onRetry(_ info: RequestInfo, attempt: Int, error: any Error, delay: Duration) {}
}

/// Hooks that do nothing.
public struct NoopHooks: HeyHooks {
    public init() {}
}

/// Several hooks as one. Start events reach them in order and end events in reverse, so spans
/// and traces nest.
public struct ChainHooks: HeyHooks {
    private let hooks: [any HeyHooks]

    public init(_ hooks: any HeyHooks...) {
        self.hooks = hooks
    }

    public init(_ hooks: [any HeyHooks]) {
        self.hooks = hooks
    }

    public func onOperationStart(_ info: OperationInfo) {
        for hook in hooks { hook.onOperationStart(info) }
    }

    public func onOperationEnd(_ info: OperationInfo, result: OperationResult) {
        for hook in hooks.reversed() { hook.onOperationEnd(info, result: result) }
    }

    public func onRequestStart(_ info: RequestInfo) {
        for hook in hooks { hook.onRequestStart(info) }
    }

    public func onRequestEnd(_ info: RequestInfo, result: RequestResult) {
        for hook in hooks.reversed() { hook.onRequestEnd(info, result: result) }
    }

    public func onRetry(_ info: RequestInfo, attempt: Int, error: any Error, delay: Duration) {
        for hook in hooks { hook.onRetry(info, attempt: attempt, error: error, delay: delay) }
    }
}

/// Hooks that print operations, and optionally requests and retries, for debugging.
public struct ConsoleHooks: HeyHooks {
    public var logOperations: Bool
    public var logRequests: Bool
    public var logRetries: Bool

    public init(logOperations: Bool = true, logRequests: Bool = false, logRetries: Bool = true) {
        self.logOperations = logOperations
        self.logRequests = logRequests
        self.logRetries = logRetries
    }

    public func onOperationStart(_ info: OperationInfo) {
        guard logOperations else { return }
        let mutation = info.isMutation ? " [mutation]" : ""
        let resource = info.resourceId.map { " #\($0)" } ?? ""
        print("[HEY] \(info.service).\(info.operation)\(resource)\(mutation)")
    }

    public func onOperationEnd(_ info: OperationInfo, result: OperationResult) {
        guard logOperations else { return }
        if let error = result.error {
            print("[HEY] \(info.service).\(info.operation) failed (\(result.duration)): \(error)")
        } else {
            print("[HEY] \(info.service).\(info.operation) completed (\(result.duration))")
        }
    }

    public func onRequestStart(_ info: RequestInfo) {
        guard logRequests else { return }
        let retry = info.attempt > 1 ? " (attempt \(info.attempt))" : ""
        print("[HEY] -> \(info.method) \(info.url)\(retry)")
    }

    public func onRequestEnd(_ info: RequestInfo, result: RequestResult) {
        guard logRequests else { return }
        let cache = result.fromCache ? " (cached)" : ""
        print("[HEY] <- \(info.method) \(info.url) \(result.statusCode) (\(result.duration))\(cache)")
    }

    public func onRetry(_ info: RequestInfo, attempt: Int, error: any Error, delay: Duration) {
        guard logRetries else { return }
        print("[HEY] Retrying \(info.method) \(info.url) (attempt \(attempt), waiting \(delay)): \(error)")
    }
}
